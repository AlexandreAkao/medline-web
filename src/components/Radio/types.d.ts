interface IRadioProps extends InputProps {
  children?: React.ReactNode;
  primary?: boolean;
}
