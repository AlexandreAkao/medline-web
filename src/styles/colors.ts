const colors = {
  primary: {
    normal: '#248F8D',
    light: '#279d9a',
    lighter: '#E9F4F4',
    darker: '#234342',
    darkest: '#257472',
  },
  white: {
    normal: '#fff',
    light: '#ccc',
    lighter: '#f0f0f0',
    lightest: '#fcfcfc',
  },
  black: {
    normal: '#000',
    soft: '#1d2129',
  },
  grey: {
    normal: '#808080',
    dark: '#3d3d3d',
  },
  blue: {
    normal: '#5041FF',
    light: '#0075FF',
  },
};

export default colors;
