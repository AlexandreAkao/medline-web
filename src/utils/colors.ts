export const getColorAlpha = (color: string, alpha = 1) => {
  const alphaLevel = Math.round(255 * alpha).toString(16);
  return color + alphaLevel;
};
