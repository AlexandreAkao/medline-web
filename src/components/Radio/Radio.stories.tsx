import { ComponentStory, ComponentMeta } from '@storybook/react';

import Radio from 'components/Radio';
import colors from 'styles/colors';

export default {
  title: 'Radio',
  args: {},
  component: Radio,
  decorators: [
    Story => (
      <div
        style={{
          background: colors.primary,
          width: '100%',
          height: '100%',
          padding: 20,
        }}
      >
        <Story />
      </div>
    ),
  ],
} as ComponentMeta<typeof Radio>;

const Template: ComponentStory<typeof Radio> = args => <Radio {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  children: 'Primary',
  id: 'primary',
  primary: true,
  onChange: event => console.log(event.target.checked),
};

export const Secondary = Template.bind({});
Secondary.args = {
  children: 'Secondary',
  id: 'secondary',
  primary: false,
};
