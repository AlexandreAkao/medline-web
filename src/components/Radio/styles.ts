import styled from 'styled-components';

import colors from 'styles/colors';

export const RadioInput = styled.input`
  height: 20px;
  width: 20px;
  cursor: pointer;
`;

export const RadioLabel = styled.label`
  margin-left: 8px;
  font-weight: 600;
  font-size: 1.125rem;
  cursor: pointer;
`;

export const RadioContainer = styled.div<IRadioProps>`
  display: inline-flex;
  justify-content: center;
  align-items: center;
  height: 20px;

  ${RadioLabel} {
    color: ${({ primary }) => (primary ? colors.black.soft : colors.white.normal)};
  }
`;
