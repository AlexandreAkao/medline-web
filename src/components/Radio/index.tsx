import { RadioContainer, RadioInput, RadioLabel } from 'components/Radio/styles';

function Radio({ id, children, primary, ...rest }: IRadioProps) {
  return (
    <RadioContainer primary={primary}>
      <RadioInput {...rest} type="radio" id={id} />
      <RadioLabel htmlFor={id}>{children}</RadioLabel>
    </RadioContainer>
  );
}

export default Radio;
