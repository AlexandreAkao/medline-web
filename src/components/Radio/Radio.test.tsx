import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import * as stories from 'components/Radio/Radio.stories';
import Radio from 'components/Radio';
import colors from 'styles/colors';

const { Primary, Secondary } = composeStories(stories);

describe('Radio', () => {
  const renderComponent = (customProps: Partial<IRadioProps> = {}) => {
    const props: IRadioProps = {
      children: 'radio-test',
      id: 'radio-test-id',
      ...customProps,
    };
    return render(<Radio {...props} />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render a primary radio button', () => {
    render(<Primary />);
    const labelElement = screen.getByText(/Primary/i);
    expect(labelElement).toBeInTheDocument();
    expect(labelElement).toHaveStyle({
      color: colors.black.soft,
    });
  });

  it('should render a secondary radio button', () => {
    render(<Secondary />);
    const labelElement = screen.getByText(/Secondary/i);
    expect(labelElement).toBeInTheDocument();
    expect(labelElement).toHaveStyle({
      color: colors.white.normal,
    });
  });

  it('should trigger function on change radio value', async () => {
    const onChange = vi.fn();
    renderComponent({ onChange });

    const radioElement = screen.getByRole('radio');
    const labelElement = screen.getByText(/Radio-test/i);

    expect(labelElement).toBeInTheDocument();
    expect(radioElement).toBeInTheDocument();
    expect(radioElement).not.toBeChecked();

    await userEvent.click(radioElement);
    expect(radioElement).toBeChecked();
    expect(onChange).toBeCalled();
  });
});
