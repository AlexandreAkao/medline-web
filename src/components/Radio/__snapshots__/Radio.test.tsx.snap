// Vitest Snapshot v1

exports[`Radio > should match with snapshot 1`] = `
<div>
  <div
    class="sc-dkzDqf eMxdaI"
  >
    <input
      class="sc-bczRLJ itSYxX"
      id="radio-test-id"
      type="radio"
    />
    <label
      class="sc-gsnTZi cTEWgn"
      for="radio-test-id"
    >
      radio-test
    </label>
  </div>
</div>
`;
